### ISSUE TITLE
[bug] tf2_ros::Buffer:transform always assumes timeout

### ISSUE THREAD

**Issue by Aposhian** [ros2/geometry2#589]
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 22.04
- Installation type:
  - binaries
- Version or commit hash:
  - humble
- DDS implementation:
  - n/a
- Client library (if applicable):
  - rclcpp

#### Steps to reproduce issue
- don't set `setUsingDedicatedThread(true)`
- try to call `transform()` on a `tf2_ros::Buffer`
- see the `tf2_ros::threading_error`

#### Expected behavior
I just want to call `transform` using the latest transform in the tf buffer.

#### Actual behavior
 `lookupTransform` allows for this, but `transform` always calls an overloaded version of `lookupTransform` that takes a timeout, which then checks for a dedicated thread. This means I get spam ERROR logs about not using a dedicated thread.

#### Additional details
This could be fixed two ways:
- removing `timeout` as default argument on `transform` and making a separate overload that takes an explicit `timeout` arg
- on `canTransform`, only check for threading if the `timeout` is non-zero.

**Comment by clalancette** [comment]
This seems like a reasonable feature request.  We'd be happy to review a PR that implements this.

**Comment by oss-vineet** [comment]
Hi @clalancette 
I can help out with this. Does using approach 2 in `buffer.cpp` look right to you? Something like:

```cpp
if (!checkAndErrorDedicatedThreadPresent(errstr) && timeout != tf2::Duration(0.0)) {
    return false;
  }
```

**Comment by Aposhian** [comment]
You will want to switch the order of those checks, since `checkAndErrorDedicatedThreadPresent` logs if it fails the check. If you check for timeout first then it will shortcut and not evaluate the check.

**Comment by oss-vineet** [comment]
> You will want to switch the order of those checks, since `checkAndErrorDedicatedThreadPresent` logs if it fails the check. If you check for timeout first then it will shortcut and not evaluate the check.

You're right about that, that's my bad. I preferred approach 2 to approach 1 because those who have implemented this code should be able to transform as is without having to pass another kind of signature. I'll write up a PR for this.

**Comment by peci1** [comment]
This can be closed as #593 fixed it.
